lib/binding/kvdb_interface.c

```diff
@@ -453,6 +439,75 @@ hse_kvdb_close(struct hse_kvdb *handle)
     return err;
 }
 
+hse_err_t
+hse_kvdb_attach(
+    const char *kvdb_home_tgt,
+    const char *kvdb_home_src,
+    const char *paths[HSE_MCLASS_COUNT])
+{
+    merr_t         err;
+    char           pidfile_path[PATH_MAX];
+    struct pidfh  *pfh_tgt = NULL, *pfh_src = NULL;
+    size_t         tlen, slen;
+
+    if (HSE_UNLIKELY(!kvdb_home_tgt || !kvdb_home_src)) {
+        log_err("KVDB attach %s home must be provided", kvdb_home_src ? "target" : "source");
+        return merr(EINVAL);
+    }
+
+    tlen = strnlen(kvdb_home_tgt, PATH_MAX);
+    slen = strnlen(kvdb_home_src, PATH_MAX);
+    if (HSE_UNLIKELY(tlen == PATH_MAX || slen == PATH_MAX)) {
+        log_err("KVDB attach %s home path length cannot be longer than PATH_MAX",
+                tlen == PATH_MAX ? "target" : "source");
+        return merr(ENAMETOOLONG);
+    }
+
+    if (HSE_UNLIKELY(tlen == 0 || slen == 0)) {
+        log_err("KVDB attach %s home must be a non-zero length path",
+                tlen == 0 ? "target" : "source");
+        return merr(EINVAL);
+    }
+
+    /* Keep both the source and target kvdb busy during the attach operation */
+    err = kvdb_home_pidfile_path_get(kvdb_home_tgt, pidfile_path, sizeof(pidfile_path));
+    if (err) {
+        log_errx("Failed to create KVDB pidfile path (%s/kvdb.pid): @@e", err, kvdb_home_tgt);
+        return err;
+    }
+
+    pfh_tgt = pidfile_open(pidfile_path, S_IRUSR | S_IWUSR, NULL);
+    if (!pfh_tgt) {
+        err = (errno == EEXIST) ? merr(EBUSY) : merr(errno);
+        log_errx("Failed to open KVDB pidfile (%s): @@e", err, pidfile_path);
+        goto out;
+    }
+
+    err = kvdb_home_pidfile_path_get(kvdb_home_src, pidfile_path, sizeof(pidfile_path));
+    if (err) {
+        log_errx("Failed to create KVDB pidfile path (%s/kvdb.pid): @@e", err, kvdb_home_src);
+        goto out;
+    }
+
+    pfh_src = pidfile_open(pidfile_path, S_IRUSR | S_IWUSR, NULL);
+    if (!pfh_src) {
+        err = (errno == EEXIST) ? merr(EBUSY) : merr(errno);
+        log_errx("Failed to open KVDB pidfile (%s): @@e", err, pidfile_path);
+        goto out;
+    }
+
+    err = ikvdb_attach(kvdb_home_tgt, kvdb_home_src, paths);
+
+out:
+    if (pidfile_remove(pfh_src) == -1 && !err)
+        err = merr(errno);
+
+    if (pidfile_remove(pfh_tgt) == -1 && !err)
+        err = merr(errno);
+
+    return err;
+}
```

Comment:
I wouldn't check for errors on pidfile_remove().  If kvdb_attach() succeeds but pidfile_remove() fails seems like you would fail unnecessarily and then caller won't know you are attached.. right?

Reply:
Good point. Also, if pidfile_remove() fails, the next pidfile_open() is capable of handling the stale kvdb.pid file.. Will fix, thanks!